Apps/FlightSoftware/Watchdog2/src/ip.c

```diff
@@ -0,0 +1,297 @@
+    }
+
+    /* TODO: double-check the ip addresses are correct */
+
+    /* verify the checksum */
+    word_stor = ip_checksum(packet, sizeof(struct ip_hdr));
+
+    /* ip checksum should be 0 or 0xffff (negative 0) for a valid packet */
+    return (word_stor == 0xffff || word_stor == 0x0) ? 0 : word_stor;
+}
+
+/**
+ * calculate UDP checksum
+ *
+ * @param udp_header: udp header *only* (no ip header part)
+ * @param data_buf: udp data buffer
+ * @param udp_packet_len: length of udp header + data part
+ * @param ip_src: source ip address
+ * @param ip_dest: destination ip address
+ */
+uint16_t udp_checksum(uint8_t *udp_header, uint8_t *data_buf, uint16_t udp_packet_len, uint32_t ip_src, uint32_t ip_dest) {
+    struct ip_pseudohdr ph;
+    uint32_t running_chksum;
+    uint16_t chk_ret;
+
+    ph.source = ip_src;
+    ph.dest = ip_dest;
+    ph.zero = 0;
+    ph.proto = 0x11;
+    ph.udp_len = udp_packet_len;
+
+    udp_packet_len = htons(udp_packet_len);
+    udp_packet_len -= sizeof(struct udp_hdr);
+
+    // hash the pseudoheader first
+    chk_ret = ip_checksum((uint8_t *)(&ph), sizeof(struct ip_pseudohdr));
+    running_chksum = ~chk_ret;
+
+    // then hash the udp header
+    chk_ret = ip_checksum(udp_header, sizeof(struct udp_hdr));
+    running_chksum += ~chk_ret;
+
+    /* check if there has been an overflow */
+    if (running_chksum >= 0xffff) {
+        /* wrap around if carry */
+        running_chksum -= 0xffff;
+    }
+
+    // finally, hash the data
+    chk_ret = ip_checksum(data_buf, udp_packet_len);
+    running_chksum += ~chk_ret;
+
+    /* check if there has been an overflow */
+    if (running_chksum >= 0xffff) {
+        /* wrap around if carry */
+        running_chksum -= 0xffff;
+    }
+
+    // return the final checksum
+    return ~running_chksum;
+}
+
+/**
+ * Send a UDP datagram
+ */
+static uint8_t ipudp_tx_buf[sizeof(struct ip_hdr) + sizeof(struct udp_hdr)];
+void ipudp_send_packet(uint8_t *data, uint16_t data_len) {
+    uint16_t out_len;
+    static uint16_t packet_nbr = 0;
+    uint16_t chksm;
+    struct ip_hdr *ip_hdr = (struct ip_hdr *)ipudp_tx_buf;
+    struct udp_hdr *udp_hdr = (struct udp_hdr *)(ip_hdr + 1);
+
+    /* we can only take a maximum of data_len - sizeof(struct ip_hdr) - sizeof(struct udp_hdr) */
+
+    /* make the ip header first */
+    ip_hdr->ver_hdrlen = 0x45;
+    ip_hdr->tos = 0;
+    ip_hdr->pckt_len = htons(data_len + sizeof(struct ip_hdr) + sizeof(struct udp_hdr));
+    ip_hdr->id = packet_nbr++;
+    ip_hdr->id = htons(ip_hdr->id);
+    ip_hdr->flgs = 0;
+    // don't really care
+    ip_hdr->ttl = 0xff;
+    // udp = 0x11
+    ip_hdr->proto = 0x11;
+    // checksum is 0 for now
+    ip_hdr->iphdr_checksum = 0;
+    // SLIP configuration payload address
+    // 192.168.103.2 => 0xC0A86702 => big endian'd
+    ip_hdr->source = 0x0267A8C0;
+    // SLIP configuration spacecraft address
+    // 192.168.103.1 => 0xC0A86701 => big endian'd
+    ip_hdr->dest = 0x0167A8C0;
```

Comment:
Nit: These are some very magic numbers, would be better to pull out these + ports as constants